aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/contexts/SegmentContextExecutors.java

```diff
@@ -0,0 +1,76 @@
+package com.amazonaws.xray.contexts;
+
+import static com.amazonaws.xray.utils.LooseValidations.checkNotNull;
+
+import com.amazonaws.xray.AWSXRay;
+import com.amazonaws.xray.AWSXRayRecorder;
+import com.amazonaws.xray.entities.Entity;
+import com.amazonaws.xray.entities.Segment;
+import java.util.concurrent.Executor;
+import javax.annotation.Nullable;
+
+/**
+ * {@link Executor}s that will mount a segment before running a command. When switching threads, for example when instrumenting an
+ * asynchronous appilcation, it is recommended to use one of these {@link Executor}s to make sure callbacks have the trace segment
+ * available.
+ *
+ * <p>For example, <pre>{@code
+ * DynamoDbAsyncClient client = DynamoDbAsyncClient.create();
+ *
+ * client.getItem(request).thenComposeAsync(response -> {
+ *     // If we did not provide the segment context executor, this request would not be traced correctly.
+ *     return client.getItem(request2);
+ * }, SegmentContextExecutors.newSegmentContextExecutor()); *
+ * }</pre>
+ */
+public final class SegmentContextExecutors {
+
+    /**
+     * Returns a new {@link Executor} which will run any tasks with the current segment mounted.
+     */
+    public static Executor newSegmentContextExecutor() {
+        return newSegmentContextExecutor(AWSXRay.getCurrentSegmentOptional().orElse(null));
+    }
+
+    /**
+     * Returns a new {@link Executor} which will run any tasks with the provided {@link Segment} mounted. If {@code segment} is
+     * {@code null}, the executor is a no-op.
+     */
+    public static Executor newSegmentContextExecutor(@Nullable Segment segment) {
+        return newSegmentContextExecutor(AWSXRay.getGlobalRecorder(), segment);
+    }
+
+    /**
+     * Returns a new {@link Executor} which will run any tasks with the provided {@link Segment} mounted in the provided
+     * {@link AWSXRayRecorder}. If {@code segment} is {@code null}, the executor is a no-op.
+     */
+    public static Executor newSegmentContextExecutor(AWSXRayRecorder recorder, @Nullable Segment segment) {
+        if (!checkNotNull(recorder, "recorder") || segment == null) {
+            return Runnable::run;
+        }
+        return new SegmentContextExecutor(recorder, segment);
+    }
+
+    private static class SegmentContextExecutor implements Executor {
+        private final AWSXRayRecorder recorder;
+        private final Segment segment;
+
+        private SegmentContextExecutor(AWSXRayRecorder recorder, Segment segment) {
+            this.recorder = recorder;
+            this.segment = segment;
+        }
+
+        @Override
+        public void execute(Runnable command) {
+            Entity previous = recorder.getTraceEntity();
+            recorder.setTraceEntity(segment);
+            try {
+                command.run();
+            } finally {
```

Comment:
I'm wary of this since it could result in double logging I think. For example, if it was a spring app I'd expect a spring exception handler to be listening to unhandled exceptions and adding to the segment appropriately. That being said, maybe we don't have any instrumentation that does this. Let me know if you still think this is a good place for such handling.

Reply:
That's a fair point. I was thinking about how we add the exception in function capturing (sub)segments [here](https://github.com/aws/aws-xray-sdk-java/blob/master/aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorder.java#L190-L200), but we also create the (sub)segments in those cases rather than using the current one in the context. Since using the segment in the context could risk double adding exceptions, e.g. for Spring segments like you mentioned, I think it's ok as is.